### PR TITLE
CoreRT support

### PR DESCRIPTION
**Pull request by adamsitnik**
There is a throughput difference for PlainText TechEmpower benchmark for CoreRT vs .NET Core 3.0 on Linux. It's big enough that the benchmarks should show where the problem is.

This PR adds CoreRT support. BenchmarkDotNet supported CoreRT for a while, but I needed to fix some minor issues (https://github.com/dotnet/BenchmarkDotNet/pull/1001) and update our scripts to support it.

@jorive CoreRT does not support full reflection, so to run the benchmarks for CoreRT we need to run the host process as .NET Core process let is use reflection etc to build a CoreRT benchmark process. This is why I needed to differentiate framework and target framework moniker in the Python scripts. When we want to run .NET Core benchmarks, framework = moniker. When we want to run the CoreRT ones, framework = CoreRT, but moniker is netcoreapp3.0.

@jkotas to run all our benchmarks for CoreRT we now just need to run following command:

```cmd
.\scripts\benchmarks_ci.py --frameworks corert --filter *
```

this script downloads .NET Core 3.0 cli from Azure, setups everything and run the benchmarks using BenchmarkDotNet. I am going to run them for Linux and share the results soon

### CODE REVIEW

**Comment by MarcoRossignoli** [comment]
@adamsitnik this https://github.com/dotnet/BenchmarkDotNet/pull/1001/files#diff-83177b9866652943fa5d5bb8f80d847aR91 could fix other issue?
However should be slower 2 fail(build/publish) before actual run.

File: scripts/micro_benchmarks.py [hunk 1]
```diff
@@ -70,13 +70,43 @@ def __get_framework_channel_map() -> dict:
         }
 
     @staticmethod
-    def get_channel(framework: str) -> str:
+    def __get_framework_target_framework_moniker_map() -> dict:
```
Comment (jorive) [inline scripts/micro_benchmarks.py]: nit: rename function to something more meaningful (`__get_corert_target_framework_mapping`)

File: scripts/micro_benchmarks.py [hunk 2]
```diff
@@ -70,13 +70,43 @@ def __get_framework_channel_map() -> dict:
...
+        return {
+            # to run CoreRT benchmarks we need to run the host BDN process as latest .NET Core
```
Comment (jorive) [inline scripts/micro_benchmarks.py]: nit: move this comment to the line above, as a function comment.

File: scripts/micro_benchmarks.py [hunk 3]
```diff
@@ -220,10 +250,40 @@ def __process_arguments(args: list) -> Tuple[list, bool]:
     parser = add_arguments(parser)
     return parser.parse_args(args)
 
+def __getBenchmarkDotNetArguments(
```
Comment (jorive) [inline scripts/micro_benchmarks.py]: nit: two lines between global function definitions.

File: scripts/micro_benchmarks.py [hunk 4]
```diff
@@ -220,10 +250,40 @@ def __process_arguments(args: list) -> Tuple[list, bool]:
...
+        args: list,
+        framework: str
+        ) -> list:
```
Comment (jorive) [inline scripts/micro_benchmarks.py]: Thanks for removing the duplicated code :)

File: scripts/benchmarks_ci.py [hunk 5]
```diff
@@ -342,7 +342,8 @@ def __run_benchview_scripts(args: list, verbose: bool) -> None:
     benchviewpy.machinedata(working_directory=bin_directory)
 
     for framework in args.frameworks:
-        buildinfo = __get_build_info(args, framework)
+        target_framework_moniker = FrameworkAction.get_target_framework_moniker(framework)
+        buildinfo = __get_build_info(args, )
```
Comment (jorive) [inline scripts/benchmarks_ci.py]: Missing second argument.

File: scripts/micro_benchmarks.py [hunk 6]
```diff
@@ -220,10 +250,40 @@ def __process_arguments(args: list) -> Tuple[list, bool]:
...
+    if args.bdn_arguments:
+        run_args += args.bdn_arguments
+        
+    # we need to tell BenchmarkDotNet where to restore the packages
+    # if we don't it's gonna restore to default global folder
+    run_args += ['--packages', get_packages_directory()]
+    # required for CoreRT where host process framework != benchmark process framework
+    run_args += ['--runtimes', framework]
+        
+    return run_args
```
Comment (jorive) [inline scripts/micro_benchmarks.py]: nit: add an extra line after function definition.

File: scripts/micro_benchmarks.py [hunk 7]
```diff
@@ -220,10 +250,40 @@ def __process_arguments(args: list) -> Tuple[list, bool]:
     parser = add_arguments(parser)
     return parser.parse_args(args)
 
+def __getBenchmarkDotNetArguments(
+        args: list,
```
Comment (jorive) [inline scripts/micro_benchmarks.py]: nit: make *args the last parameter.
```python
def __getBenchmarkDotNetArguments(framework: str, *args)
```

File: scripts/benchmarks_ci.py (hunk 5 above)
Comment (adamsitnik) [reply scripts/benchmarks_ci.py]: good catch! and a noobish bug on my side

File: scripts/micro_benchmarks.py (hunk 3 above)
Comment (adamsitnik) [reply scripts/micro_benchmarks.py]: this convention is crazy, but ok ;p

File: scripts/micro_benchmarks.py (hunk 1 above)
Comment (adamsitnik) [reply scripts/micro_benchmarks.py]: I decided to remove it and replace entire method with one if statement.

File: scripts/benchmarks_ci.py (hunk 5 above)
Comment (jorive) [reply scripts/benchmarks_ci.py]: I recommend using `pyflakes` when modifying the python scripts, and, ideally, open an issue to create a CI that run linters on this code.

File: scripts/micro_benchmarks.py (hunk 3 above)
Comment (jorive) [reply scripts/micro_benchmarks.py]: [PEP 8 -- Style Guide for Python Code](https://www.python.org/dev/peps/pep-0008/)

Comment (jorive) [reply scripts/micro_benchmarks.py]: You can use linters to help on this too ;)

File: scripts/micro_benchmarks.py [hunk 8]
```diff
@@ -251,9 +243,10 @@ def __process_arguments(args: list) -> Tuple[list, bool]:
     return parser.parse_args(args)
 
 def __getBenchmarkDotNetArguments(
-        args: list,
-        framework: str
+        framework: str,
+        args: list
```
[the remark this answers is not in the document]
Comment (adamsitnik) [reply scripts/micro_benchmarks.py]: @jorive If I do so I get following error:

```log
AttributeError: 'tuple' object has no attribute 'category'
```

[the remark this answers is not in the document]
Comment (jorive) [reply scripts/micro_benchmarks.py]: I think that in the caller site, you need to pass `args` instead of `*args`

**Comment by adamsitnik** [comment]
@jorive I updated BDN version to include https://github.com/dotnet/BenchmarkDotNet/pull/1013

thanks to @wojtpl2 BDN is now smarter, and does same thing as the Python script (dotnet restore, dotnet build --no-restore, dotnet run --no-build --no-restore) and does NOT lock the files

I have tested that for .NET Core 3.0 (regular toolchain), CoreRun and CoreRT

I believe that this PR is ready to be merged ;)

**Comment by jorive** [comment]
:shipit:

**Pull request merged**